src/tls/openssl/tls.c

```diff
@@ -396,6 +406,101 @@ int tls_set_certificate(struct tls *tls, const char *pem, size_t len)
 	return err;
 }
 
+/**
+ * Set the certificate and private key on a TLS context
+ *
+ * @param tls      TLS Context
+ * @param cert     Certificate in DER format
+ * @param key_type Private key type
+ * @param len_cert Length of certificate DER bytes
+ * @param key      Private key in DER format, will be read from cert if NULL
+ * @param len_key  Length of private key DER bytes
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int tls_set_certificate_der(struct tls *tls, enum tls_key_type key_type,
+		const uint8_t *cert, size_t len_cert, const uint8_t *key,
+		size_t len_key)
+{
+	const uint8_t *buf_cert, *buf_key;
+	X509 *x509 = NULL;
+	EVP_PKEY *pkey = NULL;
+	int r, type, err = ENOMEM;
+
+	if (!tls || !cert || !len_cert || (key && !len_key))
+		return EINVAL;
+
+	switch (key_type) {
+		case TLS_KEY_TYPE_EC:
+			type = EVP_PKEY_EC;
+			break;
+		case TLS_KEY_TYPE_RSA:
+			type = EVP_PKEY_RSA;
+			break;
+		default:
+			return EINVAL;
+	}
+
+	buf_cert = cert;
+
+	x509 = d2i_X509(NULL, &buf_cert, len_cert);
+	if (!x509)
+		goto out;
+	if (!key) {
+		buf_key = buf_cert;
+		len_key = len_cert - (buf_cert - cert);
+	}
+	else {
+		buf_key = key;
```

Comment:
I'm aware that the pointer is updated, but this is not an issue in our use case.